jni/src/common/Foreign/JNI/Unsafe.hs

```diff
@@ -1059,3 +1068,40 @@ getDirectBufferCapacity (upcast -> jbuffer) = do
       return capacity
     else
       throwIO DirectBufferFailed
+
+data GlobalRefCleaner o = GlobalRefCleaner (IORef [o]) (MVar Bool) (Async ())
```

Comment:
I don't think `o` will be ever instantiated with any other types than java references. The following should help with type-checking:
```
-data GlobalRefCleaner o = GlobalRefCleaner (IORef [o]) (MVar Bool) (Async ())
+data GlobalRefCleaner = GlobalRefCleaner (IORef [JObject]) (MVar Bool) (Async ())
```
There is an `upcast` function that can convert any `J ty` value into a `JObject`.

Also, this definition will be better documented if it uses record syntax to name the fields of `GlobalRefCleaner`. From the top of my head, some possible names are `nextBatchToDelete`, `nextAction` and `cleanerAsync`.

Instead of using a `Bool` for the `MVar`, consider using a type `data CleanerAction = Terminate | DeleteReferences`, which makes explicit what the boolean values would mean.